### ISSUE TITLE
`#[dtor]` does not run on macOS 12

### ISSUE THREAD

**Issue by BertalanD** [mmastrac/rust-ctor#240]
Sometime between April 2018 and macOS 12's release, `__mod_term_func` was deprecated and `dyld` doesn't call functions stored in it anymore. The new expected behavior is to register global destructors via `__cxa_atexit` from a global constructor. 

This is how `__attribute__((destructor))` is lowered in Xcode 14:
```c
// test.c
#include <stdio.h>

__attribute__((destructor)) void test() {
    puts("bye");
}
```

```asm
; clang -S test.c
	.section	__TEXT,__text,regular,pure_instructions
	.build_version macos, 13, 0	sdk_version 13, 0
	.globl	_test 
	.p2align	2
_test: 
	adrp	x0, l_str@PAGE
	add	x0, x0, l_str@PAGEOFF
	b	_puts

	.section	__TEXT,__StaticInit,regular,pure_instructions
	.p2align	2
___GLOBAL_init_65535:
	adrp	x0, _test@PAGE
	add	x0, x0, _test@PAGEOFF
	adrp	x2, ___dso_handle@PAGE
	add	x2, x2, ___dso_handle@PAGEOFF
	mov	x1, #0
	b	___cxa_atexit

	.section	__DATA,__mod_init_func,mod_init_funcs
	.p2align	3
	.quad	___GLOBAL_init_65535
	.section	__TEXT,__cstring,cstring_literals
l_str:
	.asciz	"bye"

.subsections_via_symbols
```

References:
- https://reviews.llvm.org/D45578
   > we are planning to  deprecate the __mod_term_funcs section on Darwin in the future.
- https://reviews.llvm.org/D121736
   > For MachO, lower `@llvm.global_dtors` into `@llvm_global_ctors` with `__cxa_atexit` calls to avoid emitting the deprecated `__mod_term_func`.
- https://github.com/gcc-mirror/gcc/commit/fabe8cc41e9b01913e2016861237d1d99d7567bf
   > For at least one target (Darwin) the platform convention is to register static destructors (i.e. __attribute__((destructor))) with __cxa_atexit rather than placing them into a list that is run by some other mechanism.

**Comment by mmastrac** [comment]
Thanks for this - it should be a pretty easy change as OSX was really the exception for not lowering to atexit in this library.

**Comment by mmastrac** [comment]
Reopening this as we should be using `__cxa_atexit`

**Comment by mmastrac** [comment]
Likely fixed properly in 28773755d21075079cefb114a4764ca21e89a568 - we now use `__cxa_atexit`

**Comment by BertalanD** [comment]
Just got around to testing the latest release on macOS Ventura. All tests pass; thank you so much for the quick fixes!
